### ISSUE TITLE
Remove @GameProperty annotation

### ISSUE THREAD

**Issue by ssoloff** [triplea-game/triplea#3250]
Now that reflective access to methods annotated with `@GameProperty` has been removed, this annotation is no longer necessary.  However, it does provide some metadata that may be useful, and, if so, that metadata should probably be captured someplace else.

The four pieces of metadata associated with this annotation are described below:

Metadata | Description
:-- | :--
`xmlProperty` | Indicates the property can be set by the `GameParser`, and that it can appear in a map XML file.
`gameProperty` | Indicates the property can be set by an `ObjectPropertyChange` or a `ChangeAttachmentChange` instance.
`adds` | Indicates the property is a collection that will be augmented each time it is set rather than being overwritten with the new values.
`virtual` | Indicates the property does not have a backing field and is only used for convenience.

`virtual` was only added to help `ValidateAttachmentsTest` better isolate which attachment properties needed to be tested.  It no longer has any value since that test has since been removed.

I never quite understood what value `xmlProperty` and `gameProperty` were supposed to provide.

All XML properties should be listed and described in the PoS2 map, so I would think we should be looking at that file as the single source of truth for what constitutes an XML property.

There's currently only six attachment properties that have `gameProperty = false`:

* `AbstractTriggerAttachment#setTrigger()` (DEPRECATED)
* `UnitAttachment#setArtillery()`
* `UnitAttachment#setArtillerySupportable()`
* `UnitAttachment#setUnitSupportCount()`
* `UnitAttachment#setIsParatroop()` (DEPRECATED)
* `UnitAttachment#setIsMechanized()` (DEPRECATED)

For those that aren't deprecated, I'm not even sure that they're _not_ set via a `Change`.  Any insight here?

I think `adds` has some value, as it changes the semantics of how one sets a property (i.e. knowing you may have to reset a property value before setting it).  I think the code makes it clear whether or not a property "adds to" or not.  However, since `adds = true` only appears on the `String` setters (i.e. those called with values read from the XML), this constraint should probably be described in the PoS2 XML.

#### TL;DR:

Without `ValidateAttachmentsTest` to check the semantics of classes that have `@GameProperty` annotations, we risk these annotations (and their metadata) diverging from the implementation of the methods to which they are applied.

Questions:

* Should `@GameProperty` be removed?
* If so, what `@GameProperty` metadata should be retained and where should it be hosted?

**Comment by ron-murhammer** [comment]
Yeah. There isn't a whole lot of value let in these and I'd be open to just removing them.

In regards to those that don't have gameProperty, my guess is that for at least the artillery ones they actually generate a support attachment under the covers and are really just an XML shortcut:
UnitAttachment#setArtillery()
UnitAttachment#setArtillerySupportable()

**Comment by Cernelius** [comment]
@ron-murhammer 
>In regards to those that don't have gameProperty, for the artillery ones they actually generate a support attachment under the covers and are really just an XML shortcut and I believe the deprecated ones essentially do the same where they set a different property and are just to keep map XML compatible.

@ron-murhammer It would be beneficial to be clearer to everyone what is deprecated.
Regarding the not deprecated ones, I believe such options should stay not deprecated, as I see they are not currently listed as deprecated, and provide an useful easier mapmaking tool for (the vast majority of) simple or traditional maps, having also the benefit or lowering the chance of map bugs (for setting attachments wrong).
I suggest having a full list of everything that it is deprecated, and that list being accessible to mapmakers lacking any programming skills.
If I'm not overlooking such an extant list, or if it is not that accessible, I encourage you (or anyone that has the right to determine what is to be considered deprecated) making one in the Wiki section at this repository. I guess it should be not a lot of work, and the mapmakers really need to be informed of anything they are supposed not to use (otherwise what is even the point of deprecating anything, in the first place).
I guess currently what is deprecated is simply removed from the PoS2 XML, but, still, having a full list of anything supported but deprecated would be better, possibly distinguishing between what is supported doing nothing and what is still functional.
Again, I surely suggest those support options not being part of such a list, even if redundant (if they would be, I guess we should update World At War, as it is bad to use deprecated elements in the moment you claim in the changelog to have updated the game to 1.9).

---

@ssoloff If I'm understanding what you are talking about (a good chance I'm not), you should not delete GameProperty, as that saves from doing pointless changes, via triggers, to:

```
UnitAttachment#setArtillery()
UnitAttachment#setArtillerySupportable()
UnitAttachment#setUnitSupportCount()
```

that are supported, as non necessary but useful (they also provide an arguably cleaner display in the tooltips) alternatives for xml coding of simple / traditional maps, and they get rewritten into supportattachment, upon initializing the game. Hence, thereafter that, they are completely useless (a mapmaker should trigger changes to the attachment they created, in case, tho, for such map, you should rather not use them, but directly define supportattachments).

---

The

`AbstractTriggerAttachment#setTrigger()`

is a very old deprecated alternative to "conditions” that have been kept not to oblige updating a bunch of existent maps (the only reason of the change is sensibility).

In here:
https://github.com/triplea-game/triplea/issues/1033
I suggested to get rid of this deprecated element, but I don't believe that was made.

> 14A) Engine only:
> option name="trigger"
> Remove this option from the engine.
> 
> 14B) Map's xml only:
> "trigger"->"conditions"

---

The

```
UnitAttachment#setIsParatroop() (DEPRECATED)
UnitAttachment#setIsMechanized() (DEPRECATED)
```

are very old disabled (supported doing nothing) and deprecated element that, under my suggestion here:
https://github.com/triplea-game/triplea/issues/1033
were supposed to have been fully deleted from the maps and the engine; in accordance to that, I assume you should fully delete all remaining occurrence of those in the engine:

---

To know what of the listed changes proposals at
https://github.com/triplea-game/triplea/issues/1033
were actually done, refer to:
https://github.com/triplea-game/triplea/wiki/Upgrade-Maps-Information

If you are able to, I suggest you running something to determine what maps currently use that deprecated "trigger" options (Total World War is one of them), as I would surely suggest (as I already did) wholly removing that confusing item from the engine, and updating all maps having it, by the way I previously quoted.
If this gets done, then, out of your list, only

```
UnitAttachment#setArtillery()
UnitAttachment#setArtillerySupportable()
UnitAttachment#setUnitSupportCount()
```

would remain; thus finally nicely cleaning up the engine, on this matter.
If that would be unfeasible, for savegame support (too many important games losing backward compatibility), I still suggest mass updating the games, giving a warning (to update and finish your saves in 1 year), and deleting that element from the engine 1 year from the mass update.

**Comment by ssoloff** [comment]
@Cernelius The proposal in this issue is to not remove the properties themselves (e.g. `artillery`, `artillerySupportable`, etc.) but just a particular piece of Java syntax we (used to) use in the code to tell us how that property was being used by the engine.  If we remove the `@GameProperty` annotations, it will have no effect on maps or on how the engine processes maps.  (Think of `@GameProperty` as nothing more than a comment for developers when looking at the code.)

**Comment by Cernelius** [comment]
@ssoloff That is not what I said. I understood you are not going to delete any of those (and I actually suggested to delete all not deprecated ones); I was just wondering if the `@GameProperty` annotation was intended to avoid triggers possibly changing those elements (which would be pointless, thus good to disable), or anyway highlighting all items that are not anymore effective past initializing and, if that is the case, my suggestion is keeping (why not keeping something that clarifies and allows detecting such particular cases, not to be mixed up with the many other regular and actual `@GameProperty`).

**Comment by ssoloff** [comment]
@Cernelius Sorry for the misunderstanding.

> I was just wondering if the `@GameProperty` annotation was intended to avoid triggers possibly changing those elements

`@GameProperty` never had any effect on engine behavior.  It was simply used during testing to alert devs if they did not implement a game property according to how the annotation said it was supposed to be implemented (it's kind of circular logic, but... :smile:).

If your concern is was the _documentation_ provided by this annotation ever used by a dev to avoid the situation you described, well, I don't know.  Based on @ron-murhammer's answer above, I'm going to assume it was not.

**Comment by Cernelius** [comment]
Ok, it's merely developers call, if nobody else is going to see that; I was wondering if that was the way to remove the option from the game, after initialize, thus assuring the game would crash, if a trigger would try to change it, thereafter (instead of changing something left to do nothing) (like in happened when "IsTwoHits", or whatever that was called, was deprecated).
In case anybody might overlook it, I want just to underline that those
```
UnitAttachment#setIsParatroop() (DEPRECATED)
UnitAttachment#setIsMechanized() (DEPRECATED)
```
should not exist anymore at all, in the engine and the maps, as per what I believe was agreed, done and documented.
